### ISSUE TITLE
Basin / Location association

### ISSUE THREAD

**Issue by adamscarberry** [USACE/cwms-data-api#609]
I think I know the answer to this question, but I'm asking to promote a discussion and standardization in the CWMS database.

_Asking the question with Access to Water in mind._

**Is there a way to get a list of basins, and then turn around and get a list of locations that are associated with that basin?**

I am unaware of a **_functional_** basins endpoint (looking at the swagger docs).  Even if that worked, the next need of actually getting a list of locations for each basin does not appear available.  Location Groups looks interesting, but I'm afraid the application of groups, names, usage, etc would vary between offices and therefore not be consistent to pull at a national level.

**Comment by adamscarberry** [comment]
Not really a part of my original question, but because CDA does not use the A2W tables to determine what timeseries paths an office has published for each location (upward reporting), there is no way to know what the official tsid is for each location.  Is that being determined by another manual assignment using CWMS-Vue?

**Comment by MikeNeilson** [comment]
Yeah, the published tsids are the ones included in the `DMZ Include List` group. Currently that's done now within CWMSVue.

For the basins. You're right I don't think there is now, but I do think it makes sense to be able to get the locations for a basin, or other appropriate meta data. 

Also i'm fairly certain the basin end point does work (maybe not for that) but few districts actually have any data configured for it. I don't believe any have been asked to do so.

**Comment by MikeNeilson** [comment]
Though that probably wouldn't be too difficult to make a solid initial guess. Most of the lat/longs are correct now.

**Comment by adamscarberry** [comment]
I don't think CWMS watersheds are stored in the database as actual geometry, unless that happened in the last few years.  I though basins were entered during one of the A2W data call spirals as just a point at the mouth of the basin.

Trying to run a spacial query for all locations inside a watershed might be a challenge without the watershed shape.

**Comment by MikeNeilson** [comment]
hmm, you're right, I know there is definitely some geospatial data, but I'm probably thinking of the district boundaries.
But we do have a spatial index on the lat/long fields so it's not like we couldn't just use a simple watershed geometry to lasso initial sets. 

Basically just throwing it out there, I think I'd rather have the districts actually just add the data. But we don't yet have the support for actually writing that anywhere. However, the changes to improve the basin endpoints will be starting "soon" (can't say more yet for legal reasons as this is a public forum) but if you have specific thoughts desires for how to use it now is the time.

E.g. Please describe what a **functional** basin end point means to you here. We'll take it into account. Might not be what we end up with for basins, which will be focused on modelling support, but we'll be able to incorporate that need somewhere.

**Comment by MikeNeilson** [comment]
@adamkorynta @rma-psmorris @rma-rripken FYI, approved for the time for initial discussion. Will shift task approval when appropriate but seems good to get the discussion going since there's interest.

**Comment by adamscarberry** [comment]
I agree that districts should probably make the associations.  Do you think the districts could use location group assignments?  This was done with the RDL (used for jasper reports) to create associations between basins and locations.  Also to associate projects with locations.  I'm not suggesting the use of RDL editor, but rather the way that it associates data in the database (uses it's own schema, but I think it uses existing location group assignment capabilities).  Can CWMS-Vue add/edit location groups in a way where locations could be added inside a group that represents a basin?  The trick is consistency across the nation.

Also I looked at the following endpoint: https://cwms-data.usace.army.mil/cwms-data/location/group?office=LRH&location-category-like=Basins

Do you know if `assigned-locations` is functional or a future feature?

**Comment by rma-rripken** [comment]
curl -X GET --location "https://cwms-data.usace.army.mil/cwms-data/location/group?office=LRH&location-category-like=Basins&include-assigned=true" \
    -H "accept: application/json" \
    -H "Cache-Control: no-cache"

**Comment by rma-rripken** [comment]
Looks like assigned-locations should have defaulted to null and not an empty list so that it wouldn't be serializing when the user didn't request the extra step of retrieving the assigned-locations.   
If you add that extra flag ("include-assigned=true") you should start seeing that field be populated.

**Comment by adamscarberry** [comment]
Thanks @rma-rripken that is interesting.  Is this query hard wired for location group "RDL_Basins"?  While this could work for LRH it won't work for the offices that didn't use RDL.

<img width="552" alt="image" src="https://github.com/USACE/cwms-data-api/assets/46494851/94aa0ab5-5afb-4260-8007-6ba478042170">

Examples where this doesn't work: 

https://cwms-data.usace.army.mil/cwms-data/location/group?office=LRN&location-category-like=Basins&include-assigned=true

https://cwms-data.usace.army.mil/cwms-data/location/group?office=SPK&location-category-like=Basins&include-assigned=true

https://cwms-data.usace.army.mil/cwms-data/location/group?office=MVP&location-category-like=Basins&include-assigned=true

<img width="592" alt="image" src="https://github.com/USACE/cwms-data-api/assets/46494851/699e4a3e-357c-411c-90cf-741629758c39">


This goes beyond a technical issue.  It's a standards problem where the offices haven't been directed to enter basins and associate the locations to those basins in a common way across the enterprise.

**Comment by MikeNeilson** [comment]
No, that just showed because the like is  case-insenstive regex like so the RDL_basin happened to match. if you did `^Basins$` it would match Basins exactly (well b or B) but as you pointed out, the districts would need to actually define the basin groups.

**Comment by rma-rripken** [comment]
Is there anything to do on this issue?  Is it worth changes LocationGroup DTO so that its possible to distinguish between "didn't fetch assigned-locations" and "fetched assigned locations and there weren't any"

**Comment by MikeNeilson** [comment]
Yeah, I think that's a good idea. I think the field being null for "not requested" and an valid but empty listed makes sense... but then again I'm a programmer.

**Comment by adamkorynta** [comment]
Shouldn't the caller know that they requested the assigned locations or not?

**Comment by MikeNeilson** [comment]
hmm, true, but it doesn't hurt to provide a response that let's you think about it.

I mean how many times have both of us been confused by a response we specifically requested?

**Comment by adamkorynta** [comment]
I think I'd still have to read the documentation to know what null vs empty collection means. I'm thinking flipping the default to true would make more sense. That way you're explicitly setting to false if you don't want all the data. I think for almost all getAll methods, query parameters filter/limit the data rather than expand it.

**Comment by adamscarberry** [comment]
Do you think it makes sense to include location `KIND` to these `assigned-locations`.  If calling from a front end, I might want to filter by project locations only (and then I still have to go fetch each one to get additional metadata).

**Comment by adamkorynta** [comment]
yeah, and I'd also want to add in location kind filters to the catalog and /locations getAll queries.

**Comment by rma-rripken** [comment]
About the location-kind - are you saying you want the returned data to include the location kind in the information about each assigned-location?  Or that you want to be able to filter out locations from the returned assigned-locations by the location-kind?  
Filter locations out of the assigned-locations field in this end-point feels weird to me.  I think if that is what is wanted/needed it makes more sense to do it in the catalog or /locations end-point.

**Comment by adamscarberry** [comment]
I would want it in the payload so I could filter it out in a web UI.  Consider you have an office landing page where you want to display the list of basins.  Clicking on each basin will give you a list of locations in that basin.  What if I wanted to show the projects at the top (or in a separate tab) and then the rest of the locations.  Having the kind would allow me to filter those out.

Is there a better endpoint for my example?

**Comment by rma-rripken** [comment]
I'd expect the location-group end-point to return detailed information about the group and enough info about the assigned-location to link the group to the location.   But I wouldn't expect the location-group end-point to returned detailed info about each location in the group.
I'd expect detailed location info from the /location end-point.   
The /catalog end-point is maybe not an obvious place to look because it doesn't  explain what its a catalog of but it can return location catalogs and ts catalogs.   
The location catalog doesn't have a kind filter but it does return the kind for each returned location. 
We should maybe add to the /location end-point and the /timeseries end-point some documentation or links to the /catalog

**Comment by rma-rripken** [comment]
The /catalog end-point when used for locations is maybe what the /location getAll end-point should have been?  idk.   The /catalog concept is powerful and useful but its maybe not very discoverable.

**Comment by adamscarberry** [comment]
This is a work in progress, but this is an example endpoint (for LRH locations) we are using for the new access to water site.  We have no watershed context at the moment (as that is not standardized across the offices).

https://a2w.rsgis.dev/cda/reporting/providers/lrh/locations

**Comment by rma-rripken** [comment]
Cool, I will check that out later today.  Love to see usages in the wild!

**Comment by adamscarberry** [comment]
We're hitting the CWMS database directly as stitching together the relationships via CDA wasn't viable.

**Comment by rma-psmorris** [comment]
Hello @adamscarberry, there are two application specific implementations of basins via location groups. RDL as you have noted (part of the Jasper Services) and REGI also has a basin implementation via location groups. Neither implementation do location kind enforcement as far as I'm aware. Both implementations were also done using location groups because there was not a generalized CWMS database basin connectivity implementation. The database implementation was done after these location group implementations and does do location kind enforcement. There is a corresponding I/O, data model, UI implementation of the CWMS basin stream connectivity in REGI, but it is a v1.0 and also not widely used. It also has an algorithm implemented (Java) for sorting all locations by stream stationing. The database connectivity also has a limited CDA implementation. I've been lobbying for basin connectivity support to be refactored into CWMSVue or some other CWMS tool as it is not specific to REGI, but that hasn't happened yet nor do I see it on the horizon currently.

All three implementations have limited usage from the districts, but if we are going to put our eggs into some basket I would think that would the generalized implementation in the database. Unfortunately that implementation does not yet have CDA coverage (I think that will be started on late spring or early summer).

**Comment by adamscarberry** [comment]
@rma-psmorris Understood. Thanks for the update.
